File: algo/lc.146.cc
#include <iostream>
#include <list>
#include <map>
#include <unordered_map> // faster on lookup time
using namespace std;

class LRUCache {
private:
    struct CacheNode {
        int key;
        int value;
        CacheNode(int k, int v) : key(k), value(v) {}
    };

    int capacity_;
    list<CacheNode> cachelist_;
    unordered_map<int, list<CacheNode>::iterator> cachemap_;

public:
    LRUCache(int capacity) {
        this->capacity_ = capacity;
    }
    
    int get(int key) {
        // not found: -1
        if (cachemap_.find(key) == cachemap_.end())
            return -1;
        // found: step1: move the node to top
        cachelist_.splice(cachelist_.begin(), cachelist_, cachemap_[key]);
        // found: step2: update map
        cachemap_[key] = cachelist_.begin();
        // found: step3: return the value
        return cachemap_[key]->value;
    }
    
    void put(int key, int value) {
        if (cachemap_.find(key) == cachemap_.end()) {
            // not found: check capacity first
            if (cachelist_.size() >= capacity_) {
                cachemap_.erase(cachelist_.back().key);
                cachelist_.pop_back();
            }
            // insert to top, add to map
            cachelist_.push_front(CacheNode(key, value));
            cachemap_[key] = cachelist_.begin();
        } else {
            // found: move to top, update map
            cachemap_[key]->value = value;
            cachelist_.splice(cachelist_.begin(), cachelist_, cachemap_[key]);
            cachemap_[key] = cachelist_.begin();
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

int
main(void)
{
	LRUCache cache = LRUCache( 2 /* capacity */ );

	cache.put(1, 1);
	cache.put(2, 2);
	cout << cache.get(1) << endl;      // returns 1
	cache.put(3, 3);    // evicts key 2
	cout << cache.get(2) << endl;       // returns -1 (not found)
	cache.put(4, 4);    // evicts key 1
	cout << cache.get(1) << endl;       // returns -1 (not found)
	cout << cache.get(3) << endl;       // returns 3
	cout << cache.get(4) << endl;       // returns 4

	return 0;
}

File: algo/lc.146.py
class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {} # Dict[int,List[int, int]]
        self.capacity = capacity
        self.timestamp = 0

    def get(self, key: int) -> int:
        # case: cache hit
        if key in self.cache:
            self.cache[key][0] = self.timestamp
            self.timestamp += 1
            return self.cache[key][-1]
        # case: not available
        else:
            return -1

    def put(self, key: int, value: int) -> None:
        # case: key does not exist and cache not full
        if key not in self.cache and len(self.cache) < self.capacity:
            self.cache[key] = [self.timestamp, value]
            self.timestamp += 1
        # case: key does not exist and cache full
        elif key not in self.cache and len(self.cache) == self.capacity:
            keyoldest = min(self.cache.keys(), key=lambda x: self.cache[x][0])
            self.cache.pop(keyoldest)
            self.cache[key] = [self.timestamp, value]
        # case: key exists
        elif key in self.cache:
            self.cache[key] = [self.timestamp, value]
        else:
            raise Exception('should not happen')
        self.timestamp += 1

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
